debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java

```diff
@@ -84,8 +84,8 @@ protected void start(BlockingConsumer<ChangeEvent> eventConsumer, Consumer<Throw
         LoggingContext.PreviousContext previousContext = taskContext.configureLoggingContext(CONTEXT_NAME);
         try {
             CompletableFuture.runAsync(this::delaySnapshotIfNeeded, executorService)
-                             .thenRun(() -> this.takeSnapshot(eventConsumer))
-                             .thenRun(() -> this.startStreaming(eventConsumer, failureConsumer))
+                             .thenRunAsync(() -> this.takeSnapshot(eventConsumer), executorService)
```

Comment:
@jpechane, can you elaborate a bit on this fix? And why it was not an issue before 0.9? And is it ensured that any exceptions during snapshotting won't be lost due to the async execution?

Reply:
@gunnarmorling It was not issue before 0.9 because the problem is present since `this::delaySnapshotIfNeeded` was introduced. We must run snapshot in a separate thread as the thread from which we start the snapshot is reading and processing events from queue to which snapshot thread is pushing. If not doen in the separatae thread then the snapshot would fill the queue and then block.
The original code has an unfortunate timing issue hidden
 `runAsync(this::delaySnapshotIfNeeded, executorService)` runs in a separate thread as expected, `.thenRun(() -> this.takeSnapshot(eventConsumer))` runs either in the thread that was started before - all works as it should - or if the thread finishes before this code is executed then it runs in the caller thread which means that queue is not consumed.
If we make a sequence of `runAsync.thenRunAsync` it is guranteed that both taskas are executed sequentially in the context of `executorService` thread.